File: web/porw-browser/delta_wasm.js
// Resident delta application. Handles own WASM allocations until the caller rewinds
// the kernel heap; no payload or genotype is retained in JavaScript.
import { decodeHeader } from "./model.js";
import {
  decodeDelta,
  decodeDelta2,
  decodeDelta3,
  isDelta2,
  isDelta3,
  deltaId,
} from "./delta.js";
import * as V from "./verify.js";
const handles = new WeakSet();
const decode = (b) =>
  isDelta3(b)
    ? decodeDelta3(b)
    : isDelta2(b)
      ? decodeDelta2(b)
      : decodeDelta(b);
function shape(h, length) {
  if (h.version !== 2) throw new Error("FLYBRAINv2 base required");
  if (
    !Number.isSafeInteger(h.neurons) ||
    h.neurons > 0xffffffff ||
    !Number.isSafeInteger(h.synapses) ||
    h.synapses > 0x7fffffff ||
    length % 4096
  )
    throw new Error("invalid base payload shape");
}
export function uploadDeltaBase(k, bytes, { baseModelId = null } = {}) {
  const mark = k.mark();
  try {
    const byteLength = bytes.byteLength;
    const hdr = decodeHeader(bytes);
    shape(hdr, byteLength);
    // A supplied identity has the same already-verified contract as applyDelta.
    if (baseModelId && baseModelId.length !== 32)
      throw new Error("baseModelId must be 32 bytes");
    const ptr = k.put(bytes);
    const rc = k.exports.porw_delta_validate(
      ptr + hdr.synOffset,
      hdr.synapses,
      hdr.neurons,
    );
    if (rc === -1)
      throw new Error("base records must be sorted by (post, pre) and unique");
    if (rc) throw new Error("base neuron index out of range");
    const modelId = baseModelId
      ? new Uint8Array(baseModelId)
      : k.merkleRoot(k.weightsLeaves(ptr, byteLength / 4096, 0));
    const h = Object.freeze({
      kernel: k,
      ptr,
      byteLength,
      hdr: Object.freeze(hdr),
      modelId,
      allocationBytes: k.mark() - mark,
    });
    handles.add(h);
    return h;
  } catch (e) {
    k.release(mark);
    throw e;
  }
}
export function applyDeltaWasm(k, base, deltaBytes, { resolve = null } = {}) {
  if (
    !handles.has(base) ||
    base.kernel !== k ||
    base.ptr + base.byteLength > k.mark()
  )
    throw new Error("invalid resident base handle or kernel");
  const mark = k.mark(),
    e = k.exports,
    h = base.hdr,
    rec = base.ptr + h.synOffset,
    n = h.synapses;
  try {
    const d = decode(deltaBytes),
      version = isDelta3(deltaBytes) ? 3 : isDelta2(deltaBytes) ? 2 : 1;
    const check = (d) => {
      if (d.neurons !== h.neurons)
        throw new Error(
          `neuron count mismatch: base ${h.neurons}, delta ${d.neurons}`,
        );
      if (!V.eq(d.baseModelId, base.modelId))
        throw new Error("base model id mismatch");
    };
    check(d);
    // The resident path writes the COMPACT materialization. An in-place cross (layout 1) keeps every base record at its
    // offset and has another model_id; producing the compact payload for it would silently load the wrong brain.
    if (version === 3 && d.layout === 1)
      throw new Error("in-place layout is not implemented in the WASM path: use applyDelta (delta.js)");
    const sample = (d, seed) => {
      const out = k.alloc(n * 4),
        scratch = k.mark(),
        rows = k.alloc(d.rTable.length * 8);
      k.u32(rows, d.rTable.length * 2).set(d.rTable.flat());
      const rc = e.porw_sample_records(
        rec,
        n,
        Number(seed & 0xffffffffn),
        Number(seed >> 32n),
        d.meanRatioQ16,
        rows,
        d.rTable.length,
        out,
      );
      k.release(scratch);
      if (rc < 0) throw new Error("WASM sampling failed: " + rc);
      return out;
    };
    const visiting = new Set(),
      cache = new Map();
    let ancestors = 0;
    const genotype = (bytes, asParent = false, depth = 0) => {
      if (depth > 64 || ++ancestors > 256)
        throw new Error("delta ancestor limit exceeded");
      const id = V.hex(deltaId(bytes)),
        is3 = isDelta3(bytes);
      if (!is3 && !isDelta2(bytes))
        throw new Error("a genotype needs a procedural delta (v2 or v3)");
      const dd = is3 ? decodeDelta3(bytes) : decodeDelta2(bytes);
      check(dd);
      if (asParent && dd.ops.length)
        throw new Error("a parent must carry no explicit ops");
      if (visiting.has(id)) throw new Error("delta ancestor cycle");
      if (cache.has(id)) return cache.get(id);
      visiting.add(id);
      let p;
      if (!is3) p = sample(dd, dd.seed);
      else {
        const parent = (pid) => {
          if (pid.every((x) => x === 0)) return 0;
          const key = V.hex(pid);
          if (visiting.has(key)) throw new Error("delta ancestor cycle");
          const b = resolve && resolve(key);
          if (!b) throw new Error(`parent delta ${key} not provided`);
          if (!V.eq(deltaId(b), pid))
            throw new Error(`parent delta ${key} hash identity mismatch`);
          return genotype(b, true, depth + 1);
        };
        const a = parent(dd.parentA),
          b = parent(dd.parentB);
        p = sample(dd, dd.seed ^ 0x4452415744524157n);
        e.porw_delta_cross(
          rec,
          n,
          a,
          b,
          p,
          Number(dd.seed & 0xffffffffn),
          Number(dd.seed >> 32n),
          dd.granularity,
          dd.mutRateQ32,
          p,
        );
      }
      visiting.delete(id);
      cache.set(id, p);
      return p;
    };
    const counts = version === 1 ? 0 : genotype(deltaBytes);
    const opPtr = k.alloc(d.ops.length * 10),
      opView = new DataView(k.memory.buffer, opPtr, d.ops.length * 10);
    d.ops.forEach((o, i) => {
      opView.setUint32(i * 10, o.pre, true);
      opView.setUint32(i * 10 + 4, o.post, true);
      opView.setInt16(i * 10 + 8, o.w, true);
    });
    const args = [
      rec,
      n,
      counts,
      version === 1 ? 0 : d.minSyn,
      opPtr,
      d.ops.length,
      version === 1 ? 1 : 0,
    ];
    const count = e.porw_delta_merge(...args, 0);
    if (count < 0) throw new Error("delete of a record the base lacks");
    const name = new TextEncoder().encode(d.name);
    if (name.length > 65535) throw new Error("name too long");
    const synOffset = 30 + name.length + h.neurons * 8,
      byteLength = Math.ceil((synOffset + count * 10) / 4096) * 4096;
    const tmp = k.alloc(byteLength);
    k.u8(tmp, byteLength).fill(0);
    const view = new DataView(k.memory.buffer, tmp, byteLength);
    k.u8(tmp, 12).set(new TextEncoder().encode("FLYBRAINv2\0\0"));
    view.setBigUint64(12, BigInt(h.neurons), true);
    view.setBigUint64(20, BigInt(count), true);
    view.setUint16(28, name.length, true);
    k.u8(tmp + 30, name.length).set(name);
    k.u8(tmp + 30 + name.length, h.neurons * 8).set(
      k.u8(base.ptr + h.neuronOffset, h.neurons * 8),
    );
    e.porw_delta_merge(...args, tmp + synOffset);
    // Move the result down over scratch before releasing it. copyWithin is memmove,
    // including for overlapping regions and shared WebAssembly memory.
    const ptr = Math.ceil(mark / 16) * 16;
    new Uint8Array(k.memory.buffer).copyWithin(ptr, tmp, tmp + byteLength);
    k.release(mark);
    const retained = k.alloc(byteLength);
    if (retained !== ptr) throw new Error("unexpected WASM alignment");
    const hdr = Object.freeze({
      name: d.name,
      neurons: h.neurons,
      synapses: count,
      synOffset,
      version: 2,
      neuronOffset: 30 + name.length,
    });
    return {
      kernel: k,
      ptr,
      byteLength,
      hdr,
      allocationBytes: k.mark() - mark,
      delta: d,
      version,
    };
  } catch (err) {
    k.release(mark);
    throw err;
  }
}

File: web/porw-browser/delta_wasm.c
#include "sample_wasm.h"
#include <stdint.h>
#define EXPORT(n) __attribute__((export_name(n)))
static uint32_t rd32(const uint8_t *p) {
  return (uint32_t)p[0] | ((uint32_t)p[1] << 8) | ((uint32_t)p[2] << 16) |
         ((uint32_t)p[3] << 24);
}
static int32_t weight(const uint8_t *p) {
  return (int16_t)((uint32_t)p[8] | ((uint32_t)p[9] << 8));
}
static int cmp(const uint8_t *a, const uint8_t *b) {
  uint32_t ap = rd32(a + 4), bp = rd32(b + 4);
  if (ap != bp)
    return ap < bp ? -1 : 1;
  ap = rd32(a);
  bp = rd32(b);
  return ap == bp ? 0 : ap < bp ? -1 : 1;
}
EXPORT("porw_delta_validate")
int32_t porw_delta_validate(const uint8_t *r, uint32_t n, uint32_t neurons) {
  for (uint32_t i = 0; i < n; i++) {
    if (rd32(r + 10 * i) >= neurons || rd32(r + 10 * i + 4) >= neurons)
      return -2;
    if (i && cmp(r + 10 * (i - 1), r + 10 * i) >= 0)
      return -1;
  }
  return 0;
}
EXPORT("porw_delta_merge")
int32_t porw_delta_merge(const uint8_t *r, uint32_t n, const uint32_t *counts,
                         uint32_t min, const uint8_t *ops, uint32_t m,
                         uint32_t strict, uint8_t *out) {
  uint32_t i = 0, j = 0, total = 0;
  while (i < n || j < m) {
    const uint8_t *a = r + 10 * i, *b = ops + 10 * j;
    int c = i == n ? 1 : j == m ? -1 : cmp(a, b);
    const uint8_t *src;
    int32_t w;
    if (c < 0) {
      src = a;
      w = weight(a);
      if (counts) {
        uint32_t v = counts[i];
        i++;
        if (v < min)
          continue;
        w = w < 0 ? -(int32_t)v : (int32_t)v;
      } else
        i++;
    } else {
      if (c > 0 && strict && weight(b) == 0)
        return -1;
      src = b;
      w = weight(b);
      j++;
      if (c == 0)
        i++;
      if (!w)
        continue;
    }
    if (out) {
      for (uint32_t x = 0; x < 8; x++)
        out[10 * total + x] = src[x];
      out[10 * total + 8] = (uint8_t)w;
      out[10 * total + 9] = (uint8_t)((uint32_t)w >> 8);
    }
    total++;
  }
  return (int32_t)total;
}
EXPORT("porw_delta_cross")
void porw_delta_cross(const uint8_t *r, uint32_t n, const uint32_t *a,
                      const uint32_t *b, const uint32_t *draw, uint32_t lo,
                      uint32_t hi, uint32_t gran, uint32_t mut, uint32_t *out) {
  for (uint32_t i = 0; i < n; i++) {
    const uint8_t *rec = r + 10 * i;
    uint32_t pre = rd32(rec), post = rd32(rec + 4);
    int32_t w = weight(rec);
    uint32_t base = w < 0 ? (uint32_t)-w : (uint32_t)w;
    uint32_t pick = (uint32_t)(porw_hash64(lo ^ 0x5049434bu, hi ^ 0x5049434bu,
                                           gran == 2 ? 0xffffffffu : pre,
                                           gran == 1 ? 0xffffffffu : post) >>
                               32);
    uint32_t v = pick >> 31 ? (b ? b[i] : base) : (a ? a[i] : base);
    if ((uint32_t)(porw_hash64(lo ^ 0x4d555421u, hi ^ 0x4d555421u, pre, post) >>
                   32) < mut)
      v = draw[i];
    out[i] = v;
  }
}
